agent_build_refactored/tools/run_in_ec2/boto3_tools.py

```diff
@@ -0,0 +1,445 @@
+            )
+
+        ssh.close()
+
+    return instance
+
+
+def create_ec2_instance(
+        boto3_session: boto3.session.Session,
+        ec2_image: EC2DistroImage,
+        name: str,
+        aws_settings: AWSSettings,
+        root_volume_size: int = None,
+):
+    """
+    Create AWS EC2 instance.
+    """
+    ec2 = boto3_session.resource("ec2")
+
+    security_groups = list(ec2.security_groups.filter(GroupNames=[aws_settings.security_group]))
+
+    if len(security_groups) != 1:
+        raise Exception(f"Number of security groups has to be 1, got '{security_groups}'")
+
+    security_group = security_groups[0]
+
+    kwargs = {}
+
+    if root_volume_size:
+        block_device_mappings = [
+            {
+                "Ebs": {
+                    "VolumeSize": root_volume_size,
+                    "DeleteOnTermination": True,
+                }
+            }
+        ]
+        kwargs.update(dict(BlockDeviceMappings=block_device_mappings))
+
+    tag_specifications = [
+            {
+                'ResourceType': 'instance',
+                'Tags': [
+                    {
+                        'Key': 'Name',
+                        'Value': name
+                    },
+                ]
+            },
+            {
+                'ResourceType': 'volume',
+                'Tags': [
+                    {
+                        'Key': 'Name',
+                        'Value': name
+                    },
+                ]
+            },
+        ]
+
+    kwargs.update(
+        dict(TagSpecifications=tag_specifications)
+    )
+
+    logger.info(f"Start new EC2 instance using image '{ec2_image.image_id}', size '{ec2_image.size_id}'")
+    attempts = 10
+    while True:
+        try:
+            instances = ec2.create_instances(
+                ImageId=ec2_image.image_id,
+                MinCount=1,
+                MaxCount=1,
+                InstanceType=ec2_image.size_id,
+                KeyName=aws_settings.private_key_name,
+                SecurityGroupIds=[security_group.id],
+                **kwargs
+            )
+            break
+        except botocore.exceptions.ClientError as e:
+            if attempts == 0:
+                logger.exception("    Giving up.")
+                raise
+
+            message = str(e)
+            # We may catch capacity limit error from AWS, so just retry.
+            no_capacity_error = f"We currently do not have sufficient {ec2_image.size_id} capacity in zones with " \
+                                f"support for 'gp2' volumes. Our system will be working on provisioning additional " \
+                                f"capacity."
+            if no_capacity_error not in message:
+                logger.exception("Unrecoverable error has occurred.")
+                raise
+
+            logger.info(f"    {e}")
+            logger.info(f"    Retry...")
+            attempts -= 1
+            time.sleep(random.randint(5, 10))
```

Comment:
Maybe sleep longer here?  I don't expect ec2 instances to be destroyed so frequently, maybe a sleep of 1 to 3 minutes is more appropriate, wdyt?